effects/anim-polygons/anim-polygons.c

```diff
@@ -0,0 +1,190 @@
+#include "2d.h"
+#include "blitter.h"
+#include "cia.h"
+#include "copper.h"
+#include "effect.h"
+#include "fx.h"
+#include "line.h"
+#include "memory.h"
+#include "pixmap.h"
+#include "types.h"
+
+/* different line drawing methods */
+#define CPULINE 0
+#define CPUEDGE 0
+#define BLITLINE 1
+
+#define WIDTH 320
+#define HEIGHT 180
+#define DEPTH 4
+
+#define HEIGHT_P1 210
+
+static BitmapT *screen;
+static CopInsT *bplptr[DEPTH];
+static CopListT *cp;
+static short active = 0;
+
+#include "anim_data.c"
+#include "anim-pal.c"
+
+/* Reading polygon data */
+static short current_frame = 0;
+static short frame_count = 260 - 5;
+static int vert_ptr = 0;
+/* Synchronization */
+int frame_diff = 0;
+int frame_sync;
+
+static void Load(void) {
+    screen = NewBitmap(WIDTH, HEIGHT, DEPTH + 1);
+    frame_sync = ReadFrameCounter();
+}
+
+static void UnLoad(void) { DeleteBitmap(screen); }
+
+static void MakeCopperList(CopListT *cp) {
+    CopInit(cp);
+    CopSetupBitplanes(cp, bplptr, screen, DEPTH);
+    {
+        short *pixels = gradient.pixels;
+        short i, j;
+
+        for (i = 0; i < HEIGHT_P1 / 10; i++) {
+            CopWait(cp, Y(i * 10 - 1), 0xde);
+            for (j = 0; j < 16; j++) CopSetColor(cp, j, *pixels++);
+        }
+    }
+    CopEnd(cp);
+}
+
+static void Init(void) {
+    EnableDMA(DMAF_BLITTER);
+    BitmapClear(screen);
+
+    SetupPlayfield(MODE_LORES, DEPTH, X(0), Y(0), WIDTH, HEIGHT);
+    cp = NewCopList(100 + gradient.height * (gradient.width + 1));
+    MakeCopperList(cp);
+    CopListActivate(cp);
+    EnableDMA(DMAF_RASTER);
+}
+
+static void Kill(void) {
+    DisableDMA(DMAF_COPPER | DMAF_RASTER | DMAF_BLITTER);
+    DeleteCopList(cp);
+}
+
+/* Get x on screen position from linear memory repr */
+static inline short CalculateX(u_short xy) {
+    return (short)(xy - (xy / WIDTH) * WIDTH);
+}
+
+/* Get y on screen position from linear memory repr */
+static inline short CalculateY(u_short xy) { return (short)(xy / WIDTH); }
+
+/* Using wrapper because it looked really messy with all those
+ * #ifs in DrawSpans. In production it won't be necessary */
+static void DrawLine(short xp, short yp, short xe, short ye) {
+#if CPULINE == 1
```

Comment:
We can safely assume Blitter provides the fastest method for drawing lines. You may remove alternatives I think.